cadcdata/cadcdata/tests/test_inventory.py

```diff
@@ -0,0 +1,571 @@
+    put_mock.assert_called_with('url1', data=ANY,
+                                headers={'Content-Type': 'text/plain',
+                                         'Content-Encoding': 'us-ascii'})
+
+    # update metadata only
+    put_mock.reset_mock()
+    post_mock = Mock()
+    basews_mock.return_value.post = post_mock
+    client._get_transfer_urls = Mock(return_value=['url1'])
+    client.cadcinfo = Mock(
+        return_value=cadcdata.FileInfo('cadc:TEST/putfile',
+                                       file_type='application/file',
+                                       encoding='none', md5sum='0x123456789'))
+    client.cadcput('cadc:TEST/putfile', file_name, replace=True,
+                   file_type='text/plain', file_encoding='us-ascii',
+                   md5_checksum='0x123456789')
+    put_mock.assert_not_called()
+    post_mock.assert_called_with('url1',
+                                 headers={'Content-Type': 'text/plain',
+                                          'Content-Encoding': 'us-ascii'})
+
+    # no update required - data and metadata identical
+    put_mock.reset_mock()
+    post_mock.reset_mock()
+    client._get_transfer_urls = Mock(return_value=['url1'])
+    client.cadcinfo = Mock(
+        return_value=cadcdata.FileInfo('cadc:TEST/putfile',
+                                       file_type='text/plain',
+                                       encoding='us-ascii',
+                                       md5sum='0x123456789'))
+    client.cadcput('cadc:TEST/putfile', file_name, replace=True,
+                   file_type='text/plain', file_encoding='us-ascii',
+                   md5_checksum='0x123456789')
+    put_mock.assert_not_called()
+    post_mock.assert_not_called()
+
+    # replace non existing file
+    client.cadcinfo.side_effect = [exceptions.NotFoundException()]
+    with pytest.raises(AttributeError):
+        client.cadcput('cadc:TEST/putfile', file_name, replace=True)
+
+    # put a new file that already exists
+    client.cadcinfo = Mock(
+        return_value=cadcdata.FileInfo('cadc:TEST/putfile',
+                                       file_type='text/plain',
+                                       encoding='us-ascii',
+                                       md5sum='0x123456789'))
+    with pytest.raises(AttributeError):
+        client.cadcput('cadc:TEST/putfile', file_name, replace=False)
+
+
+@patch('cadcdata.core.net.BaseWsClient')
+def test_remove(basews_mock):
+    client = StorageInventoryClient(auth.Subject())
+    # test a put
+    remove_mock = Mock()
+    basews_mock.return_value.delete = remove_mock
+    with pytest.raises(exceptions.UnauthorizedException):
+        client.cadcremove('cadc:TEST/removefile')
+    client._cadc_client.subject.anon = False  # authenticate the user
+
+    client._get_transfer_urls = Mock(return_value=['url1'])
+    client.cadcremove('cadc:TEST/removefile')
+    remove_mock.assert_called_with('url1')
```

Comment:
Possible test - attempt to delete a non-existent file?

Reply:
Added